californium-core/src/main/java/org/eclipse/californium/core/network/Matcher.java

```diff
@@ -174,6 +174,10 @@ public void sendRequest(Exchange exchange,final  Request request) {
 		if (request.getToken() == null) {
 			idByToken = createUnusedToken();
 			request.setToken(idByToken.token);
+			// if original request has no token set it too.
+			// this is the case where request and currentRequest is not the same
+			if (exchange.getRequest() != null && exchange.getRequest().getToken() == null)
```

Comment:
Because the API allow you to create request without token and the token is dynamically created in the matcher, but in this case the blockwise request is created in blockwiselayer before the original one go through the matcher.
```
 * +-----------------------+
 * | MessageDeliverer      |
 * +-------------A---------+
 *               |
 *             * A            
 * +-Endpoint--+-A---------+
 * |           v A         |
 * |           v A         |
 * | +---------v-+-------+ |
 * | | Stack Top         | |
 * | +-------------------+ |
 * | |  ObserveLayer     | |
 * | +-------------------+ |
 * | |  BlockwiseLayer   | |
 * | +-------------------+ |
 * | |  ReliabilityLayer | |
 * | +-------------------+ |
 * | | Stack Bottom      | |
 * | +--------+-+--------+ |
 * |          v A          |
 * |          v A          |
 * |       Matcher         |
 * |          v A          |
 * |    MessageInterceptor |  
 * |          v A          |
 * |          v A          |
 * | +--------v-+--------+ |
 * +-|    Connector      |-+
 *   +--------+-A--------+
 *            v A
 *            v A
 *         (Network)
```